torch/csrc/jit/fusion_compiler.cpp

```diff
@@ -1047,67 +1132,407 @@ struct CPUFusionFunction : public CompiledFusionFunction {
   }
-  return it->second;
+
+  return {map_size};
 }
 
-std::shared_ptr<CompiledFusionFunction> FusionCompiler::getOrCompile(Node* fusion_group) {
-  auto & graph = *fusion_group->g(attr::Subgraph);
-  AnnotatedGraph agraph(graph, fusion_group->i(attr::device));
-  for(auto & input : graph.inputs()) {
-    auto t = input->type()->expect<TensorType>();
-    agraph.input_desc.emplace_back(t);
+// See Note [Run-time shape checking code] for more explanation on the algorithm.
+at::optional<std::vector<int64_t>> FusedKernelCache::canRunKernel(at::TensorList args) {
+  AT_CHECK(args.size() == input_chunks.size(),
+           "Expected ", input_chunks.size(), " arguments, but got ", args.size());
+
+  at::optional<std::vector<int64_t>> map_size;
+  for (const auto & broadcast_group : input_broadcast_groups) {
+    if (!map_size) {
+      map_size = getMapSize(args, broadcast_group);
+      if (!map_size) {
+        return at::nullopt;
+      }
+    } else {
+      auto group_map_size = getMapSize(args, broadcast_group);
+      // NB: this checks that group_map_size is defined AND equal to map_size
+      if (map_size != group_map_size) {
+        return at::nullopt;
+      }
+    }
   }
-  for(auto & output : graph.outputs()) {
-    auto t = output->type()->expect<TensorType>();
-    agraph.output_desc.emplace_back(t);
+  return map_size;
+}
+
+void FusedKernelCache::runFallback(Stack& stack) {
+  InterpreterState(fallback_code).runOneStage(stack);
+}
+
+// NB: args are mutated in this call. map_size is mutated too, but is restored to its original
+// value before this function returns (it's an optimization).
+void FusedKernelCache::expandArgs(std::vector<at::Tensor>& args, std::vector<int64_t>& map_size) {
+  for (size_t i = 0; i < args.size(); ++i) {
+    auto & arg = args[i];
+    auto & pdesc = input_chunks[i];
+    if (pdesc.nSubtensors == 1) {
+      if (arg.sizes().equals(map_size)) continue;
+      arg = arg.expand(map_size);
+    } else {
+      map_size.at(pdesc.dim) *= pdesc.nSubtensors;
+      if (!arg.sizes().equals(map_size)) {
+        arg = arg.expand(map_size);
+      }
+      map_size.at(pdesc.dim) /= pdesc.nSubtensors;
+    }
   }
-  return getOrCompile(agraph);
 }
 
+std::unique_ptr<FusedKernel> FusedKernelCache::compileSpec(
+      const FusedKernelArgSpec& spec, const std::vector<int64_t>& map_size) {
+  AnnotatedGraph agraph {*graph, device};
 
-std::shared_ptr<CompiledFusionFunction> FusionCompiler::getOrCompile(Graph & graph,
-                                                     int device,
-                                                     at::ArrayRef<at::Tensor> inputs,
-                                                     at::ArrayRef<at::Tensor> outputs) {
-  AnnotatedGraph agraph(graph, device);
-  for(auto & i : inputs) {
-   agraph.input_desc.emplace_back(i);
+  agraph.input_desc = spec.descs();
+  at::optional<at::ScalarType> scalar_type;
```

Comment:
This seems fragile. It is making a previously unmade assumption that the fuser won't want to mix integer and floating point types. It should at least be documented here. The other parts of codegen that cannot emit integers or mixed floats/doubles are minor and will likely be fixed by someone, and this seems like a trap when making that fix. 

Is it not possible to just take the type as-is from the output Value? If we determined it was safe to fuser (ie. it was floating point), then we checked that it was floating point, meaning the type was known.

Reply:
> It is making a previously unmade assumption

We are currently making an assumption that the fuser only does `float` math, so I can't see how this is a problem. Having all inputs being integral types will trigger an assert failure below, so someone will likely find out about this issue on time. I can add a comment though.